File: tuesday/morning/claude_exercise/quadratic_map.c
#include <stdio.h>
#include <stdlib.h>

/* Quadratic map function: g(x) = 4x(1-x) */
double quadratic_map(double x) {
    return 4.0 * x * (1.0 - x);
}

/* Generate trajectory data for time series */
void generate_trajectory(double x0, int n, const char *filename) {
    FILE *fp = fopen(filename, "w");
    if (fp == NULL) {
        fprintf(stderr, "Error opening file %s\n", filename);
        return;
    }

    double x = x0;
    fprintf(fp, "t,x\n");

    for (int t = 0; t <= n; t++) {
        fprintf(fp, "%d,%.15f\n", t, x);
        x = quadratic_map(x);
    }

    fclose(fp);
}

/* Generate data for plotting the function itself */
void generate_function_data(const char *filename) {
    FILE *fp = fopen(filename, "w");
    if (fp == NULL) {
        fprintf(stderr, "Error opening file %s\n", filename);
        return;
    }

    fprintf(fp, "x,g_x\n");

    int num_points = 1000;
    for (int i = 0; i <= num_points; i++) {
        double x = (double)i / num_points;
        double g_x = quadratic_map(x);
        fprintf(fp, "%.15f,%.15f\n", x, g_x);
    }

    fclose(fp);
}

/* Generate trajectory data for histogram (long time series) */
void generate_histogram_data(double x0, int n, const char *filename) {
    FILE *fp = fopen(filename, "w");
    if (fp == NULL) {
        fprintf(stderr, "Error opening file %s\n", filename);
        return;
    }

    double x = x0;
    fprintf(fp, "x\n");

    for (int t = 0; t <= n; t++) {
        fprintf(fp, "%.15f\n", x);
        x = quadratic_map(x);
    }

    fclose(fp);
}

int main() {
    printf("Generating quadratic map data...\n");

    /* Generate function data for Figure 1 */
    generate_function_data("quadratic_function.csv");
    printf("Generated quadratic_function.csv\n");

    /* Generate trajectory for Figure 2 (x0 = 0.3, n = 150 for visualization) */
    generate_trajectory(0.3, 150, "trajectory.csv");
    printf("Generated trajectory.csv\n");

    /* Generate long trajectory for Figure 3 (x0 = 0.3, n = 100000 for histogram) */
    generate_histogram_data(0.3, 100000, "histogram_data.csv");
    printf("Generated histogram_data.csv\n");

    printf("All data files generated successfully!\n");

    return 0;
}

File: tuesday/morning/claude_exercise/quadratic_map.f90
program quadratic_map
    implicit none

    ! Quadratic map: g(x) = 4x(1-x)

    ! Main program
    call generate_function_data()
    call generate_trajectory(0.3d0, 150, 'trajectory_fortran.csv')
    call generate_histogram_data(0.3d0, 100000, 'histogram_data_fortran.csv')

    print *, 'All Fortran data files generated successfully!'

contains

    ! Quadratic map function: g(x) = 4x(1-x)
    subroutine generate_function_data()
        integer, parameter :: num_points = 1000
        real(8) :: x, g_x
        integer :: i, unit_num

        open(newunit=unit_num, file='quadratic_function_fortran.csv', status='replace')
        write(unit_num, '(A)') 'x,g_x'

        do i = 0, num_points
            x = real(i, 8) / real(num_points, 8)
            g_x = quadratic_map_func(x)
            write(unit_num, '(F20.15,A,F20.15)') x, ',', g_x
        end do

        close(unit_num)
        print *, 'Generated quadratic_function_fortran.csv'
    end subroutine generate_function_data

    ! Generate trajectory data for time series
    subroutine generate_trajectory(x0, n, filename)
        real(8), intent(in) :: x0
        integer, intent(in) :: n
        character(len=*), intent(in) :: filename
        real(8) :: x
        integer :: t, unit_num

        open(newunit=unit_num, file=filename, status='replace')
        write(unit_num, '(A)') 't,x'

        x = x0
        do t = 0, n
            write(unit_num, '(I0,A,F20.15)') t, ',', x
            x = quadratic_map_func(x)
        end do

        close(unit_num)
        print *, 'Generated ', trim(filename)
    end subroutine generate_trajectory

    ! Generate trajectory data for histogram (long time series)
    subroutine generate_histogram_data(x0, n, filename)
        real(8), intent(in) :: x0
        integer, intent(in) :: n
        character(len=*), intent(in) :: filename
        real(8) :: x
        integer :: t, unit_num

        open(newunit=unit_num, file=filename, status='replace')
        write(unit_num, '(A)') 'x'

        x = x0
        do t = 0, n
            write(unit_num, '(F20.15)') x
            x = quadratic_map_func(x)
        end do

        close(unit_num)
        print *, 'Generated ', trim(filename)
    end subroutine generate_histogram_data

end program quadratic_map
